wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketApiClient.java

```diff
@@ -0,0 +1,217 @@
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          }
+        });
+  }
+
+  /**
+   * Returns the scopes of the OAuth token.
+   *
+   * @param authenticationToken The OAuth token to inspect.
+   * @return Array of scopes from the supplied token, empty array if no scope.
+   * @throws ScmItemNotFoundException
+   * @throws ScmCommunicationException
+   * @throws ScmBadRequestException
+   */
+  public String[] getTokenScopes(String authenticationToken)
+      throws ScmItemNotFoundException, ScmCommunicationException, ScmBadRequestException {
+    final URI uri = apiServerUrl.resolve("user");
+    HttpRequest request = buildBitbucketApiRequest(uri, authenticationToken);
+    LOG.trace("executeRequest={}", request);
+    return executeRequest(
+        httpClient,
+        request,
+        response -> {
+          Optional<String> scopes = response.headers().firstValue(BITBUCKET_OAUTH_SCOPES_HEADER);
+          return Splitter.on(',')
+              .trimResults()
+              .omitEmptyStrings()
+              .splitToList(scopes.orElse(""))
+              .toArray(String[]::new);
+        });
+  }
+
+  private HttpRequest buildBitbucketApiRequest(URI uri, String authenticationToken) {
+    return HttpRequest.newBuilder(uri)
+        .headers("Authorization", "Bearer " + authenticationToken, "Accept", "application/json")
+        .timeout(DEFAULT_HTTP_TIMEOUT)
+        .build();
+  }
+
+  private <T> T executeRequest(
+      HttpClient httpClient,
+      HttpRequest request,
+      Function<HttpResponse<InputStream>, T> responseConverter)
+      throws ScmBadRequestException, ScmItemNotFoundException, ScmCommunicationException {
+    try {
+      HttpResponse<InputStream> response =
+          httpClient.send(request, HttpResponse.BodyHandlers.ofInputStream());
+      LOG.trace("executeRequest={} response {}", request, response.statusCode());
+      if (response.statusCode() == HTTP_OK) {
+        return responseConverter.apply(response);
+      } else if (response.statusCode() == HTTP_NO_CONTENT) {
+        return null;
+      } else {
+        String body = CharStreams.toString(new InputStreamReader(response.body(), Charsets.UTF_8));
+        switch (response.statusCode()) {
+          case HTTP_BAD_REQUEST:
+            throw new ScmBadRequestException(body);
+          case HTTP_NOT_FOUND:
+            throw new ScmItemNotFoundException(body);
+          default:
+            throw new ScmCommunicationException(
+                "Unexpected status code " + response.statusCode() + " " + response.toString());
+        }
+      }
+    } catch (IOException | InterruptedException | UncheckedIOException e) {
+      throw new ScmCommunicationException(e.getMessage(), e);
+    }
+  }
+
+  /**
+   * Checks if the provided url belongs to this client (Bitbucket)
+   *
+   * @param scmServerUrl the SCM url to verify
+   * @return If the provided url is recognized by the current client
```

Comment:
@return {@code true} if the provided url is recognized by the current client

Reply:
done